helper/testhelpers/docker/testhelpers.go

```diff
@@ -370,3 +373,239 @@ func copyToContainer(ctx context.Context, dapi *client.Client, containerID, from
+var _ RunCmdOpt = (*RunCmdUser)(nil)
+
+func (u RunCmdUser) Apply(cfg *types.ExecConfig) error {
+	cfg.User = string(u)
+	return nil
+}
+
+func (d *Runner) RunCmdWithOutput(ctx context.Context, container string, cmd []string, opts ...RunCmdOpt) ([]byte, []byte, int, error) {
+	runCfg := types.ExecConfig{
+		AttachStdout: true,
+		AttachStderr: true,
+		Cmd:          cmd,
+	}
+
+	for index, opt := range opts {
+		if err := opt.Apply(&runCfg); err != nil {
+			return nil, nil, -1, fmt.Errorf("error applying option (%d / %v): %v", index, opt, err)
+		}
+	}
+
+	ret, err := d.DockerAPI.ContainerExecCreate(ctx, container, runCfg)
+	if err != nil {
+		return nil, nil, -1, fmt.Errorf("error creating execution environment: %v\ncfg: %v\n", err, runCfg)
+	}
+
+	resp, err := d.DockerAPI.ContainerExecAttach(ctx, ret.ID, types.ExecStartCheck{})
+	if err != nil {
+		return nil, nil, -1, fmt.Errorf("error attaching to command execution: %v\ncfg: %v\nret: %v\n", err, runCfg, ret)
+	}
+	defer resp.Close()
+
+	var stdoutB bytes.Buffer
+	var stderrB bytes.Buffer
+	if _, err := stdcopy.StdCopy(&stdoutB, &stderrB, resp.Reader); err != nil {
+		return nil, nil, -1, fmt.Errorf("error reading command output: %v", err)
+	}
+
+	stdout := stdoutB.Bytes()
+	stderr := stderrB.Bytes()
+
+	// Fetch return code.
+	info, err := d.DockerAPI.ContainerExecInspect(ctx, ret.ID)
+	if err != nil {
+		return stdout, stderr, -1, fmt.Errorf("error reading command exit code: %v", err)
+	}
+
+	return stdout, stderr, info.ExitCode, nil
+}
+
+func (d *Runner) RunCmdInBackground(ctx context.Context, container string, cmd []string, opts ...RunCmdOpt) (string, error) {
+	runCfg := types.ExecConfig{
+		AttachStdout: true,
+		AttachStderr: true,
+		Cmd:          cmd,
+	}
+
+	for index, opt := range opts {
+		if err := opt.Apply(&runCfg); err != nil {
+			return "", fmt.Errorf("error applying option (%d / %v): %v", index, opt, err)
+		}
+	}
+
+	ret, err := d.DockerAPI.ContainerExecCreate(ctx, container, runCfg)
+	if err != nil {
+		return "", fmt.Errorf("error creating execution environment: %v\ncfg: %v\n", err, runCfg)
+	}
+
+	err = d.DockerAPI.ContainerExecStart(ctx, ret.ID, types.ExecStartCheck{})
+	if err != nil {
+		return "", fmt.Errorf("error starting command execution: %v\ncfg: %v\nret: %v\n", err, runCfg, ret)
+	}
+
+	return ret.ID, nil
+}
+
+// Mapping of path->contents
+type PathContents interface {
+	UpdateHeader(header *tar.Header) error
+	Get() ([]byte, error)
+}
+
+type FileContents struct {
+	Data []byte
+	Mode int64
+}
+
+func (b FileContents) UpdateHeader(header *tar.Header) error {
+	header.Mode = b.Mode
+	return nil
+}
+
+func (b FileContents) Get() ([]byte, error) {
+	return b.Data, nil
+}
+
+func PathContentsFromBytes(data []byte) PathContents {
+	return FileContents{
+		Data: data,
+		Mode: 0o644,
+	}
+}
+
+type BuildContext map[string]PathContents
+
+func NewBuildContext() BuildContext {
+	return BuildContext{}
+}
+
+func (ctx *BuildContext) ToTarball() (io.Reader, error) {
```

Comment:
This is a silly gripe but I don't like seeing ctx used for non-context.Context types.  It's just hardwired into my brain enough that I find it hurts readability.